libkbfs/prefetcher.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"io"
+	"sort"
+	"sync"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultIndirectPointerPrefetchCount int = 20
+	fileIndirectBlockPrefetchPriority   int = -100
+	dirEntryPrefetchPriority            int = -200
+)
+
+type prefetchRequest struct {
+	priority int
+	kmd      KeyMetadata
+	ptr      BlockPointer
+	block    Block
+}
+
+// blockRetriever specifies a method for retrieving blocks asynchronously
+type blockRetriever interface {
+	Request(ctx context.Context, priority int, kmd KeyMetadata, ptr BlockPointer, block Block, lifetime BlockCacheLifetime) <-chan error
+}
+
+type blockPrefetcher struct {
+	retriever  blockRetriever
+	progressCh chan prefetchRequest
+	shutdownCh chan struct{}
+	doneCh     chan struct{}
+}
+
+var _ Prefetcher = (*blockPrefetcher)(nil)
+
+func newBlockPrefetcher(retriever blockRetriever) *blockPrefetcher {
+	p := &blockPrefetcher{
+		retriever:  retriever,
+		progressCh: make(chan prefetchRequest),
+		shutdownCh: make(chan struct{}),
+		doneCh:     make(chan struct{}),
+	}
+	go p.run()
+	return p
+}
+
+func (p *blockPrefetcher) run() {
+	var wg sync.WaitGroup
+	defer func() {
+		wg.Wait()
+		close(p.doneCh)
+	}()
+	for {
+		select {
+		case req := <-p.progressCh:
+			if p.retriever == nil {
+				continue
+			}
+			ctx, cancel := context.WithCancel(context.Background())
+			ch := p.retriever.Request(ctx, req.priority, req.kmd, req.ptr, req.block, TransientEntry)
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				select {
+				case _ = <-ch:
+				case <-p.shutdownCh:
+					// Cancel but still wait so p.doneCh accurately represents
+					// whether we still have requests pending.
+					cancel()
+					<-ch
+				}
+			}()
+		case <-p.shutdownCh:
+			return
+		}
+	}
+}
+
+func (p *blockPrefetcher) request(priority int, kmd KeyMetadata, ptr BlockPointer, block Block) error {
+	// TODO: plumb through config, and then check the pointer and config data
+	// versions before prefetching anything. Use
+	// folderBlockOps.checkDataVersion as a template.
+	select {
+	case p.progressCh <- prefetchRequest{priority, kmd, ptr, block}:
+		return nil
+	case <-p.shutdownCh:
+		return io.EOF
```

Comment:
may be worth defining a separate shutdown error type, or doing `errors.Wrap(io.EOF)` so we know where this error comes from